src/common/transformations/src/transformations/symbolic_transformations/dereshape_matmul.cpp

```diff
@@ -0,0 +1,336 @@
+    ov::NodeVector dims;
+    // batch dims here stand for MatMul batch dims -- leaving two last dims for Matrix Multiplication
+    size_t num_batch_dims = batch_dims_source.get_partial_shape().size() - 2;
+    get_dims(batch_dims_source, 0, num_batch_dims, copy_rt_info_from, dims);
+
+    size_t non_batch_dims_start = non_batch_dims_source.get_partial_shape().size() - 2;
+    get_dims(non_batch_dims_source, non_batch_dims_start, non_batch_dims_start + 2, copy_rt_info_from, dims);
+
+    size_t num_non_const_nodes = 0;  // candidates for becoming a Constant -1 -- special value for Reshape pattern
+    for (size_t curr_i = 0; curr_i + 1 < dims.size(); ++curr_i) {
+        auto curr_node = dims[curr_i], next_node = dims[curr_i + 1];
+        bool curr_is_const = ov::op::util::is_constant(curr_node), next_is_const = ov::op::util::is_constant(next_node);
+        if (num_non_const_nodes == 0 && !curr_is_const && next_is_const) {
+            curr_node = ov::op::v0::Constant::create(ov::element::i64, {1}, {-1});
+            curr_is_const = true;
+            num_non_const_nodes += 1;
+        }
+        if (num_non_const_nodes == 0 && !next_is_const && curr_is_const) {
+            next_node = ov::op::v0::Constant::create(ov::element::i64, {1}, {-1});
+            next_is_const = true;
+            num_non_const_nodes += 1;
+        }
+        if (curr_is_const && next_is_const) {
+            dims[curr_i] = nullptr;
+            dims[curr_i + 1] = ov::op::util::make_try_fold<ov::op::v0::Concat>(ov::NodeVector{curr_node, next_node}, 0);
+        }
+    }
+    dims.erase(std::remove_if(dims.begin(),
+                              dims.end(),
+                              [](const std::shared_ptr<ov::Node>& node) {
+                                  return node == nullptr;
+                              }),
+               dims.end());
+    auto target_shape = ov::op::util::make_try_fold<ov::op::v0::Concat>(dims, 0);
+    ov::copy_runtime_info(copy_rt_info_from, target_shape);
+    return target_shape->output(0);
+}
+
+void pull_reshape_through_optional_concat_and_bea(const ov::pass::pattern::PatternValueMap& vm,
+                                                  std::shared_ptr<ov::Node> concat_label,
+                                                  std::shared_ptr<ov::Node> bea_label,
+                                                  ov::Output<ov::Node> reshape_output,
+                                                  ov::Input<ov::Node> matmul_input,
+                                                  std::vector<ov::Node*>& nodes_for_revalidation) {
+    // Reshape -- [Concat] -- [BEA with scalar] -- > MatMul
+    auto original_reshape = reshape_output.get_node_shared_ptr();
+    if (vm.count(concat_label)) {
+        auto concat_node = ov::as_type_ptr<ov::op::v0::Concat>(vm.at(concat_label).get_node_shared_ptr());
+        OPENVINO_ASSERT(concat_node != nullptr,
```

Comment:
usually transformations do not use ASSERT 
As I understand, the main idea is to continue execution without applying the transformation, so maybe it's better to use `return false;` and stop the transformation